broker.go

```diff
@@ -0,0 +1,584 @@
+package legs
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	dt "github.com/filecoin-project/go-data-transfer"
+	"github.com/filecoin-project/go-legs/p2p/protocol/head"
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-graphsync"
+	"github.com/ipld/go-ipld-prime"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	"github.com/ipld/go-ipld-prime/traversal/selector"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+)
+
+// defaultAddrTTL is the default amount of time that addresses discovered from
+// pubsub messages will remain in the peerstore.  This is twice the default
+// provider poll interval.
+const defaultAddrTTL = 48 * time.Hour
+
+var ErrSourceNotAllowed = errors.New("message source not allowed by policy")
+
+// PeerPolicyHandler is the function signature of a function given to LegBroker
+// that the broker uses to determine whether to allow or reject messages
+// originating from peer passed into the function.  Returning true indicates
+// that messages from the peer are allowed, and false indicated they should be
+// rejected.  Returning an error indicates that there was a problem evaluating
+// the peer, and results in the messages being rejected.
+type PeerPolicyHandler func(peerID peer.ID) (bool, error)
+
+// LegBroker is a message broker for pubsub messages.  LegBroker creates a
+// single pubsub subscriber that receives messages from a gossip pubsub topic,
+// and creates a stateful message handler for each message source peer.  An
+// optional external policy determines whether to allow or deny handling
+// messages from specific peers.
+//
+// Messages to separate peers are handled concurrently, and multiple messages
+// to the same peer are handled serially.  If a handler is busy handling a
+// message, and more messages arrive for the same peer, then the last message
+// replaces the previous unhandled message to avoid having to maintain queues
+// of messages.  Handlers do not have persistent goroutines, but start a new
+// goroutine to handle a single message.
+type LegBroker struct {
```

Comment:
I was trying to follow the naming already present but would like to change this.  I think I can drop the `Leg` prefix as that stutters given that the package is named `legs`.

I named it Broker because it receives messages from various peers over pubsub and then relays the messages to separate the handler for each peer.

Other names considered:
Dispatcher, Receiver, Router, Communicator

Something called `legs.Broker` seems like the most fun.